### PR TITLE
Fix: also get from translog when query matching on `_id`

### PR DESCRIPTION
**Pull request by matriv**
## Summary of the changes / Why this improves CrateDB

Previously, we were only searching if the symbols in the `WHERE` clause
were matching all the columns comprising the table's `PRIMARY KEY`, but
we didn't search if the query of the `WHERE` clause was trying on `_id`,
which if an explicit `PRIMARY KEY` is defined, is a concaneted string of
those pk column values.

Fixes: #13057

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by matriv** [comment]
Not sure if we want to address this, or just clarify in docs.
```
testFilterOn_IdOnPartitionedTableDoesNotResultInDocKeys
testFilterOn_IdDoesNotProduceDocKeysIfTableHasOnlyAClusteredByDefinition
```
will  return `docKeys()` with this.

File: server/src/main/java/io/crate/planner/WhereClauseOptimizer.java [hunk 1]
```diff
@@ -212,16 +215,10 @@ private static List<ColumnIdent> pkColsInclVersioning(DocTableInfo table,
...
-            return pkCols;
+            return Collections.unmodifiableList(Lists2.concat(table.primaryKey(), DocSysColumns.VERSION));
         } else if (seqNoAndPrimaryTermInQuery) {
-            ArrayList<ColumnIdent> pkCols = new ArrayList<>(table.primaryKey().size() + 1);
-            pkCols.addAll(table.primaryKey());
-            pkCols.add(DocSysColumns.SEQ_NO);
-            pkCols.add(DocSysColumns.PRIMARY_TERM);
-            return pkCols;
+            Collections.unmodifiableList(
+                Lists2.concat(table.primaryKey(), List.of(DocSysColumns.SEQ_NO, DocSysColumns.PRIMARY_TERM)));
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/WhereClauseOptimizer.java:221]: Why the change here? I think the old version is more efficient as it avoids an intermediate list creation

File: server/src/main/java/io/crate/planner/WhereClauseOptimizer.java [hunk 2]
```diff
@@ -161,6 +161,9 @@ public static DetailedQuery optimize(EvaluatingNormalizer normalizer,
 
         EqualityExtractor eqExtractor = new EqualityExtractor(normalizer);
         List<List<Symbol>> pkValues = eqExtractor.extractExactMatches(pkCols, query, txnCtx);
+        if (pkValues == null) {
+            pkValues = eqExtractor.extractExactMatches(List.of(DocSysColumns.ID), query, txnCtx);
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/WhereClauseOptimizer.java:165]: I wonder if there's a way to speed this up by re-using parts of the first pass?

If I remember right the EqualityExtractor tends to show up in planner flamegraphs when profiling

File: server/src/main/java/io/crate/planner/WhereClauseOptimizer.java (hunk 1 above)
Comment (matriv) [reply server/src/main/java/io/crate/planner/WhereClauseOptimizer.java:221]: You're right, was not intending to keep that, because of the `List.of()`.

File: server/src/main/java/io/crate/planner/WhereClauseOptimizer.java (hunk 2 above)
Comment (matriv) [reply server/src/main/java/io/crate/planner/WhereClauseOptimizer.java:165]: On the top of my head, and since to return `pkValues` here the query must match all of the pk columns - so we cannot just add `_id` to the `table.primaryKey()`, I could try to pass a List<List<ColumnIdent>> and have an extra iteration when matching query symbols. wdyt?

Comment (mfussenegger) [reply server/src/main/java/io/crate/planner/WhereClauseOptimizer.java:165]: >  I could try to pass a List<List> and have an extra iteration when matching query symbols. wdyt?

Sounds good. I'm not sure if it is possible, but if we can re-use some of the "content preparation" that's happening within the EqualityExtractor it would be nice.

**Comment by matriv** [comment]
Apologies @mfussenegger please don't check yet, I will request the review once I solve an issue that popped up.

**Comment by matriv** [comment]
I cannot find a way to call the extractor only once, in the first commit (approach).
With this commit some tests are failing, for example: `TransportSQLActionTest#test_primary_key_lookup_on_multiple_columns()`

If you have an idea to make the 1st approach work, please let me know.

File: server/src/test/java/io/crate/planner/SelectPlannerTest.java [hunk 3]
```diff
@@ -867,11 +867,13 @@ public void testGlobalAggregateOn3TableJoinWithNoMatch() throws Exception {
...
-            .addTable(
+            .addPartitionedTable(
                 "create table t_pk_part_generated (" +
                 "   ts timestamp with time zone," +
                 "   p as date_trunc('day', ts)," +
-                "   primary key (ts, p))")
+                "   primary key (ts, p)) " +
+                "partitioned by (p)",
+            "p")
             .build();
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/planner/SelectPlannerTest.java:877]: Why the change here?

**Review by mfussenegger — COMMENTED** [review]
Looks good overall. Left a minor suggestion (feel free to ignore), but also a question about a test change.

File: server/src/main/java/io/crate/planner/WhereClauseOptimizer.java [hunk 4]
```diff
@@ -192,6 +177,26 @@ public static DetailedQuery optimize(EvaluatingNormalizer normalizer,
                 }
             }
         }
+        int clusterIdxWithinPK = table.primaryKey().indexOf(table.clusteredBy());
+        final DocKeys docKeys;
+        if (pkValues == null && shouldUseDocKeys(table, clusteredBy, partitionValues)) {
+            pkValues = eqExtractor.extractExactMatches(List.of(DocSysColumns.ID), query, txnCtx);
+        }
```
[the remark this answers is not in the document]
Comment (matriv) [reply server/src/main/java/io/crate/planner/WhereClauseOptimizer.java:184]: i don't mind at all, I just wanted to have a `shouldUseDocKeys` and then negate it to make it more readable, method is not necessary, thx!

File: server/src/test/java/io/crate/planner/SelectPlannerTest.java (hunk 3 above)
Comment (matriv) [reply server/src/test/java/io/crate/planner/SelectPlannerTest.java]: no reason, I was confused with the test title, and thought it wanted to check partitioned table with generated column, reverted.

**Comment by matriv** [comment]
irrelevant: `org.elasticsearch.indices.recovery.IndexRecoveryIT.testTransientErrorsDuringRecoveryAreRetried` failure

**Comment by matriv** [comment]
retest this please

**Review by seut — COMMENTED** [review]
Isn't this missing a test case to validate the fix? (select by _id without refresh)

**Comment by matriv** [comment]
retest this please

**Comment by matriv** [comment]
> Isn't this missing a test case to validate the fix? (select by _id without refresh)

Yes, it was, sorry about that, I was focused on not break anything and forgot to add them

**Review by seut — APPROVED** [review]
👍

File: server/src/test/java/io/crate/integrationtests/WherePKIntegrationTest.java [hunk 5]
```diff
@@ -212,6 +212,35 @@ public void testSelectByIdWithPartitionsUsesSearch() throws Exception {
...
+        execute("create table users (name string)" +
+                "clustered by (name) with (refresh_interval = -1, number_of_replicas = 0)");
+
+        execute("insert into users values ('hoschi'), ('galoschi'), ('x')");
+
+        execute("select _id from users");
+        for (Object[] row : response.rows()) {
+            execute("select name from users where _id=?", row);
+            assertThat(response.rowCount(), is(1L));
+        }
```
[the remark this answers is not in the document]
Comment (matriv) [reply server/src/test/java/io/crate/integrationtests/WherePKIntegrationTest.java]: It is, of course! Thank you!
